### PR TITLE
Add QR support + BIP21

### PR DESCRIPTION
**Pull request by JSKitty**
## Abstract

This PR adds a fully reusable QR module in to MPW: it's first integrated usecase being for Payment Requests, MPW now supports two kinds of QR payment requests: plain address, and BIP21, for the widest coverage ensuring MPW has compatibility with most mainstream wallets and merchants / payment gateways utilising BIP21.

image

image
*Ignore the 'Recording' icon in the right, this is not a part of the MPW UI.*

The QR module being independent and re-usable also means we can re-use this functionality incredibly easily for other usecases: like encrypted QR-based wallet exports and imports.

## What does this PR address?
It addresses the problem of having to copy/paste addresses manually by text, now, one can utilise the QR codes in both MyPIVXWallet and PIVX Core for seamless, fast transfers.

## What features or improvements were added?
A fully modular QR code reader, as well as BIP21 Payment Requests support.

## How does this benefit users?
Seamless, painless payments with minimal clicks and typing. Just a camera and some PIV.

### CODE REVIEW

File: scripts/misc.js [hunk 1]
```diff
@@ -135,6 +135,40 @@ export function createQR(strData = '', domImg, size = 4) {
     domImg.firstChild.style.borderRadius = '8px';
 }
 
+/**
+ * Attempt to safely parse a BIP21 Payment Request
+ * @param {string} strReq - BIP21 Payment Request string
+ * @returns {object | false}
+ */
+export function parseBIP21Request(strReq) {
+    // Format should match: pivx:addr[?amount=x&label=x]
```
Comment (Duddino) [inline scripts/misc.js:144]: We should probably check that the address contains the `pivx:` prefix

File: scripts/scanner.js [hunk 2]
```diff
@@ -0,0 +1,75 @@
...
+            {
+                returnDetailedScanResult: true,
+                preferredCamera: 'environment',
+                highlightCodeOutline: true,
+                highlightScanRegion: true,
+            }
+        );
+
+        // Attempt to start scanning, then display the UI
+        try {
```
Comment (Duddino) [inline scripts/scanner.js:43]: I think try blocks don't work on promises that you don't await, we should use `.catch()` instead

**Review by Duddino — CHANGES_REQUESTED** [review]

**Review by Duddino — DISMISSED** [review]
tACK really like how this works

**Comment by JSKitty** [comment]
Good for re-review: Shield addresses are supported now, and conflicts fixed.

File: scripts/misc.js [hunk 3]
```diff
@@ -135,6 +136,43 @@ export function createQR(strData = '', domImg, size = 4) {
...
+    const strAddress = addressPart.substring(BIP21_PREFIX.length + 1); // remove 'pivx:' prefix
+    let cOptions = {};
+
+    // Ensure the address is valid
+    if (
+        // Standard address
+        (strAddress.length !== 34 ||
+            !cChainParams.current.PUBKEY_PREFIX.includes(strAddress[0])) &&
+        // Shield address
+        !isValidBech32(strAddress).valid
```
Comment (panleone) [inline scripts/misc.js:160]: then here take in input the prefix (ps mainnet ptest testnet)

File: scripts/misc.js [hunk 4]
```diff
@@ -153,6 +191,25 @@ export async function generateMnPrivkey() {
...
+ * @property {boolean} valid - If the string is a valid bech32 address
+ * @property {object} res - The results of the bech32 decoding
+ */
+
+/**
+ * A safe bech32 wrapper for quickly checking if an address is valid
+ * @param {string} str - Bech32 Address
+ * @returns {Bech32Check} - Both the validity and decoding results
+ */
+export function isValidBech32(str) {
```
Comment (panleone) [inline scripts/misc.js:205]: add prefix as parameter

File: scripts/misc.js [hunk 5]
```diff
@@ -153,6 +191,25 @@ export async function generateMnPrivkey() {
...
+    try {
+        return { valid: true, res: bech32.decode(str) };
```
Comment (panleone) [inline scripts/misc.js:207]: check that the prefix matches the one found by bech32

File: scripts/global.js [hunk 6]
```diff
@@ -409,6 +420,51 @@ export function selectMaxBalance(domValueInput, fCold = false) {
...
+    // Plain address (Length and prefix matches)
+    if (
+        cScan.data.length === 34 &&
+        cChainParams.current.PUBKEY_PREFIX.includes(cScan.data[0])
+    ) {
+        return guiPreparePayment(cScan.data);
+    }
+
+    // Shield address (Valid bech32 string)
+    if (isValidBech32(cScan.data).valid) {
```
Comment (panleone) [inline scripts/global.js:442]: same here take in input the prefix

**Review by panleone — COMMENTED** [review]
some comments

**Review by BreadJS — DISMISSED** [review]

**Review by panleone — DISMISSED** [review]
tACK f7e0639050db2ed6cac07af2735e9275589824e9
Sorry for the late review, finally found some time to actually test this PR!
Working as expected both for shielding and transparent + my review has been solved on discord

**Review by Liquid369 — DISMISSED** [review]
tACK f7e0639050db2ed6cac07af2735e9275589824e9
Works great, even with shitty phone cameras!

**Comment by JSKitty** [comment]
Conflicts resolved, pls review.

image

**Review by panleone — APPROVED** [review]
retack

**Pull request merged**